src/game/client/shadow/w3dvolumetricshadow.cpp

```diff
@@ -0,0 +1,512 @@
+                    // found a shared edge
+                    for (int edge = 0; edge < 3; ++edge) {
+                        if (m_polyNeighbors[i].neighbor[edge].neighborIndex == -1) {
+                            m_polyNeighbors[i].neighbor[edge].neighborIndex = (short)j;
+                            m_polyNeighbors[i].neighbor[edge].neighborEdgeIndex[0] = (short)index1;
+                            m_polyNeighbors[i].neighbor[edge].neighborEdgeIndex[1] = (short)index2;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+bool W3DShadowGeometryMesh::Allocate_Neighbors(int num_polys)
+{
+    captainslog_assert(m_numPolyNeighbors == 0);
+    captainslog_assert(m_polyNeighbors == nullptr);
+
+    m_polyNeighbors = new PolyNeighbor[num_polys];
+
+    if (m_polyNeighbors == nullptr) {
+        captainslog_assert(0);
+        return false;
+    }
+
+    m_numPolyNeighbors = num_polys;
+    return true;
+}
+
+void W3DShadowGeometryMesh::Delete_Neighbors()
+{
+
+    if (m_polyNeighbors != nullptr) {
+        delete[] m_polyNeighbors;
+        m_polyNeighbors = nullptr;
+        m_numPolyNeighbors = 0;
+    }
+}
+
+Vector3 *W3DShadowGeometryMesh::Get_Polygon_Normal(int index) const
+{
+    captainslog_assert(m_polygonNormals != nullptr);
+
+    return &m_polygonNormals[index];
+}
+
+void W3DShadowGeometryMesh::Build_Polygon_Normals()
+{
+    if (m_polygonNormals == nullptr) {
+        Vector3 *normals = new Vector3[m_numPolygons];
+
+        for (int i = 0; i < m_numPolygons; ++i) {
+            Get_Polygon_Normal(i, &normals[i]);
+        }
+
+        m_polygonNormals = normals;
+    }
+}
+
+void W3DShadowGeometryMesh::Get_Polygon_Normal(int index, Vector3 *normal) const
+{
+
+    if (m_polygonNormals != nullptr) {
+        *normal = m_polygonNormals[index];
+    }
+
+    short index_list[3];
+    Get_Polygon_Index(index, index_list);
+
+    Vector3 *v0 = Get_Vertex(index_list[0]);
+    Vector3 *v1 = Get_Vertex(index_list[1]);
+    Vector3 *v2 = Get_Vertex(index_list[2]);
+
+    Vector3 b = *v1 - *v0;
+    Vector3 a = *v1 - *v2;
+
+    Vector3::Normalized_Cross_Product(a, b, normal);
+}
+
+void W3DShadowGeometryMesh::Get_Polygon_Index(int polygon_index, short *index_list) const
+{
+    const TriIndex *v = &m_polygons[polygon_index];
+    index_list[0] = m_parentVerts[v->I];
+    index_list[1] = m_parentVerts[v->J];
+    index_list[2] = m_parentVerts[v->K];
+}
+
+W3DShadowGeometry::W3DShadowGeometry() : m_name{}, m_meshCount(0), m_numTotalsVerts(0) {}
+
+void W3DShadowGeometry::Set_Name(const char *name)
+{
+    strlcpy(m_name, name, sizeof(m_name));
+}
+
+int W3DShadowGeometry::Init(RenderObjClass *robj)
+{
+    return 1;
+}
+
+int W3DShadowGeometry::Init_From_HLOD(RenderObjClass *robj)
+{
+    HLodClass *hlod = static_cast<HLodClass *>(robj);
+
+    int lod_index = hlod->Get_LOD_Count() - 1;
+
+    W3DShadowGeometryMesh *geo_mesh = &m_meshList[m_meshCount];
```

Comment:
Urgh this is so dangerous. The L274 assert should be up here. We are possibly getting a pointer to invalid memory at this point. I would also suggest that this be changed into a reference and possibly into a function so all access's are safe. `Init_From_Mesh` for example is **not** checking if it has exhausted the mesh count!

```cpp
W3DShadowGeometryMesh& Get_Next_Mesh(){
    captainslog_dbgassert(m_meshCount < MAX_SHADOW_CASTER_MESHES, "Too many shadow sub-meshes"); 
    return m_meshList[m_meshCount];
}
```

Reply:
If it goes into bug fix territory, I would prefer it to be a separate commit for visibility (which can be part of this pull request of course).